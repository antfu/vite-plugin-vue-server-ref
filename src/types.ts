import type { Ref } from 'vue'

export interface ServerRefOptions<T extends Record<string, unknown>> {
  state?: T
  /**
   * Milliseconds to debounce the server ref updates.
   *
   * @default 10
   */
  debounce?: number
  /**
   * Default value for the client when the data is not presented.
   *
   * @default () => undefined
   */
  defaultValue?: (key: string) => unknown
  /**
   * Log info on the client
   */
  debug?: boolean
  /**
   * The vue entry name for the generated client code.
   *
   * @default 'vue'
   */
  clientVue?: string
  /**
   * Callback on server data change
   */
  onChanged?: <K extends keyof T>(name: K, data: T[K], patch: Partial<T[K]> | undefined, timestamp: number) => void
}

export interface ServerRefExtend<T> {
  $syncDown: boolean
  $syncUp: boolean
  $onChange: (fn: (data: T) => void) => void
  $onPatch: (fn: (patch: Partial<T>) => void) => void
  $patch: (patch: Partial<T>) => Promise<boolean>
}

export type ServerRef<T> = Ref<T> & ServerRefExtend<T>
export type ServerReactive<T> = T & ServerRefExtend<T>
