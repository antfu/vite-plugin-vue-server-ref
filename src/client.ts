export { klona as clone } from 'klona/json'
export { diff, apply } from './diff'

export const UNDEFINED = '__UNDEFINED___'

export function stringify(data: any) {
  return JSON.stringify(data, (key, value) => value === undefined ? UNDEFINED : value)
}

export function parse(json: string) {
  return JSON.parse(json, (key, value) => value === UNDEFINED ? undefined : value)
}

export function randId() {
  return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(2, 10)
}

export function isObject(obj: any) {
  return typeof obj === 'object' && obj !== null
}

export function reactiveSet(target: any, value: any) {
  const originalKeys = new Set(Object.keys(target))
  Object.keys(value).forEach(i => originalKeys.delete(i))
  Object.assign(target, value)
  for (const key of originalKeys)
    delete target[key]
}

export function define(target: any, key: string, value: any) {
  Object.defineProperty(target, key, {
    value,
    enumerable: false,
  })
}
