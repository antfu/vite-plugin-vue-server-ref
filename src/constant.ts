export const URL_PREFIX = ['/@server-ref/', '/@server-reactive/']
export const VIRTUAL_PREFIX = ['server-ref:', 'server-reactive:']
export const WS_EVENT = 'vue-server-ref'

export const PREFIXES = [...URL_PREFIX, ...VIRTUAL_PREFIX]
