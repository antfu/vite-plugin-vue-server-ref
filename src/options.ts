import { ServerRefOptions } from '../dist'

export const defaultOptions: Required<ServerRefOptions<any>> = {
  state: {},
  debounce: 10,
  debug: false,
  defaultValue: () => undefined,
  clientVue: 'vue',
  onChanged: () => { },
}
export function resolveOptions(options: ServerRefOptions<any>) {
  return Object.assign({}, defaultOptions, options)
}
