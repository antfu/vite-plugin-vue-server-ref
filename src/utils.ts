import type { Connect } from 'vite'

export function getBodyJson(req: Connect.IncomingMessage) {
  return new Promise<any>((resolve, reject) => {
    let body = ''
    req.on('data', chunk => body += chunk)
    req.on('error', reject)
    req.on('end', () => {
      try {
        resolve(JSON.parse(body) || {})
      }
      catch (e) {
        reject(e)
      }
    })
  })
}
