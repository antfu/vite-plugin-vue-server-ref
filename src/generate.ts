import type { ServerRefOptions } from './types'
import { WS_EVENT } from './constant'
import type { ParsedId } from './utils'
import { get } from './utils'

export function genCode(
  options: Required<ServerRefOptions<any>>,
  id: ParsedId,
) {
  const { state, defaultValue, clientVue, debug, debounce } = options
  const { key, type, prefix, diff } = id
  const access = type === 'ref' ? 'data.value' : 'data'

  return `
import { ${type}, watch } from "${clientVue}"
import { randId, stringify, parse, define, apply,${type === 'reactive' ? ' reactiveSet,' : ''}${diff ? ' clone, diff' : ''} } from "vite-plugin-vue-server-ref/client"

const data = ${type}(${JSON.stringify(get(state, key) ?? defaultValue(key))})

let onSet = []
let onPatch = []

define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))

if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  ${diff ? 'let copy = null' : ''}

  function post(payload) {
    ${debug ? `console.log("[server-ref] [${key}] outgoing", payload)` : ''}
    return fetch('${prefix + key}', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }

  ${diff
    ? `
  function makeClone() {
    copy = clone(${access})
  }

  function getDiff() {
    return diff(${access}, copy)
  }
  `
    : ''}

  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger) {
      onPatch.forEach(fn => fn(patch))
      ${debug ? `console.log("[server-ref] [${key}] patch incoming", patch)` : ''}
    }
    apply(${access}, patch)
    ${diff ? 'makeClone()' : ''}
    skipWatch = false
  }

  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    ${type === 'ref' ? 'data.value = payload.data' : 'reactiveSet(data, newData)'}
    ${debug ? `console.log("[server-ref] [${key}] incoming", newData)` : ''}
    ${diff ? 'makeClone()' : ''}
    skipWatch = false
  }
 
  ${debug ? `console.log("[server-ref] [${key}] ref", data)` : ''}
  ${debug ? `console.log("[server-ref] [${key}] initial", ${access})` : ''}
 
  import.meta.hot.on("${WS_EVENT}", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== "${key}" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })

  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })

  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return

    timer = setTimeout(()=>{
      post({
        source: id,
        ${diff ? 'patch: getDiff()' : `data: ${access}`},
        timestamp: Date.now(),
      })
    }, ${debounce})
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}

export default data
`.replace(/\n\s*\n/gm, '\n')
}
