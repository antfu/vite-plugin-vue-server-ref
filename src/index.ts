import type { Plugin } from 'vite'
import { ServerRefOptions } from './types'
import { getBodyJson } from './utils'

export * from './types'

const URL_PREFIX = '/@server-ref/'
const VIRTUAL_PREFIX = 'server-ref:'
const WS_EVENT = 'server-ref'

export default function VitePluginServerRef(options: ServerRefOptions<any> = {}): Plugin {
  const {
    state = {},
    debounce = 10,
    debug = false,
    defaultValue = () => undefined,
    clientVue = 'vue',
  } = options

  return <Plugin>{
    name: 'vite-plugin-vue-server-ref',
    resolveId(id) {
      if (id.startsWith(VIRTUAL_PREFIX))
        return URL_PREFIX + id.slice(VIRTUAL_PREFIX.length)
      return id.startsWith(URL_PREFIX)
        ? id
        : null
    },
    configureServer(server) {
      server.middlewares.use(async(req, res, next) => {
        if (!req.url?.startsWith(URL_PREFIX) || req.method !== 'POST')
          return next()

        const key = req.url.slice(URL_PREFIX.length)
        const { data, timestamp, patch } = await getBodyJson(req)

        const module = server.moduleGraph.getModuleById(URL_PREFIX + key)
        if (module)
          server.moduleGraph.invalidateModule(module)

        state[key] = data

        server.ws.send({
          type: 'custom',
          event: WS_EVENT,
          data: {
            key,
            data,
            patch,
            timestamp,
          },
        })

        options.onChanged?.(key, data, patch, timestamp)

        res.write('')
        res.end()
      })
    },
    load(id) {
      if (!id.startsWith(URL_PREFIX))
        return null

      const key = id.slice(URL_PREFIX.length)
      return `
import { ref, watch } from "${clientVue}"

const data = ref(${JSON.stringify(state[key] ?? defaultValue(key))})

data.syncUp = true
data.syncDown = true
data.paused = false
data.defer = false

if (import.meta.hot) {
  ${debug ? `console.log("[server-ref] [${key}] ref", data)` : ''}
  ${debug ? `console.log("[server-ref] [${key}] initial", data.value)` : ''}

  let skipNext = false
  let timer = null
  import.meta.hot.on("${WS_EVENT}", (payload) => {
    if (!data.syncDown || data.paused)
      return
    if (payload.key !== "${key}")
      return
    skipNext = true
    data.value = payload.data
    ${debug ? `console.log("[server-ref] [${key}] incoming", payload.data)` : ''}
  })
  watch(data, (v) => {
    if (!data.syncUp || data.paused)
      return
    if (skipNext) {
      skipNext = false
      return
    }
    if (timer)
      clearTimeout(timer)

    timer = setTimeout(()=>{
      ${debug ? `console.log("[server-ref] [${key}] outgoing", data.value)` : ''}
      fetch('${URL_PREFIX + key}', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          data: data.value,
          timestamp: Date.now(),
        })
      })
    }, ${debounce})
  }, { flush: 'sync', deep: true })
}

export default data
`
    },
  }
}
