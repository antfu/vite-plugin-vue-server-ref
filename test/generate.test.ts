import { genCode } from '../src/generate'
import { resolveOptions } from '../src/options'
import { parseId } from '../src/utils'

describe('generate', () => {
  const options = resolveOptions({})
  it('ref', () => {
    const id = parseId('server-ref:foo')!
    expect(genCode(options, id)).toMatchSnapshot()
  })

  it('deep ref', () => {
    const id = parseId('server-ref:foo/bar')!
    expect(genCode(options, id)).toMatchSnapshot()
  })

  it('reactive', () => {
    const id = parseId('server-reactive:foo')!
    expect(genCode(options, id)).toMatchSnapshot()
  })

  it('reactive deferred', () => {
    const id = parseId('server-reactive:foo?defer')!
    expect(genCode(options, id)).toMatchSnapshot()
  })
})
