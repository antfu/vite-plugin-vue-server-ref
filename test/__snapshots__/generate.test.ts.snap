// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate deep ref 1`] = `
"
import { ref, watch } from \\"vue\\"
import { randId, stringify, parse, define, apply, } from \\"vite-plugin-vue-server-ref/client\\"
const data = ref(undefined)
let onSet = []
let onPatch = []
define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))
if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  function post(payload) {
    return fetch('server-ref:foo.bar', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }
  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger)
      onPatch.forEach(fn => fn(patch))
    apply(data.value, patch)
    skipWatch = false
  }
  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    data.value = payload.data
    skipWatch = false
  }
  import.meta.hot.on(\\"vue-server-ref\\", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== \\"foo.bar\\" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })
  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })
  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return
    timer = setTimeout(()=>{
      post({
        source: id,
        data: data.value,
        timestamp: Date.now(),
      })
    }, 10)
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}
export default data
"
`;

exports[`generate reactive 1`] = `
"
import { reactive, watch } from \\"vue\\"
import { randId, stringify, parse, define, apply, reactiveSet, } from \\"vite-plugin-vue-server-ref/client\\"
const data = reactive(undefined)
let onSet = []
let onPatch = []
define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))
if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  function post(payload) {
    return fetch('server-reactive:foo', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }
  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger)
      onPatch.forEach(fn => fn(patch))
    apply(data, patch)
    skipWatch = false
  }
  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    reactiveSet(data, newData)
    skipWatch = false
  }
  import.meta.hot.on(\\"vue-server-ref\\", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== \\"foo\\" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })
  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })
  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return
    timer = setTimeout(()=>{
      post({
        source: id,
        data: data,
        timestamp: Date.now(),
      })
    }, 10)
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}
export default data
"
`;

exports[`generate reactive deferred 1`] = `
"
import { reactive, watch } from \\"vue\\"
import { randId, stringify, parse, define, apply, reactiveSet, clone, diff } from \\"vite-plugin-vue-server-ref/client\\"
const data = reactive(undefined)
let onSet = []
let onPatch = []
define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))
if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  let copy = null
  function post(payload) {
    return fetch('server-reactive:foo', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }
  function makeClone() {
    copy = clone(data)
  }
  function getDiff() {
    return diff(data, copy)
  }
  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger)
      onPatch.forEach(fn => fn(patch))
    apply(data, patch)
    makeClone()
    skipWatch = false
  }
  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    reactiveSet(data, newData)
    makeClone()
    skipWatch = false
  }
  import.meta.hot.on(\\"vue-server-ref\\", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== \\"foo\\" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })
  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })
  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return
    timer = setTimeout(()=>{
      post({
        source: id,
        patch: getDiff(),
        timestamp: Date.now(),
      })
    }, 10)
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}
export default data
"
`;

exports[`generate ref 1`] = `
"
import { ref, watch } from \\"vue\\"
import { randId, stringify, parse, define, apply, } from \\"vite-plugin-vue-server-ref/client\\"
const data = ref(undefined)
let onSet = []
let onPatch = []
define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))
if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  function post(payload) {
    return fetch('server-ref:foo', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }
  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger)
      onPatch.forEach(fn => fn(patch))
    apply(data.value, patch)
    skipWatch = false
  }
  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    data.value = payload.data
    skipWatch = false
  }
  import.meta.hot.on(\\"vue-server-ref\\", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== \\"foo\\" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })
  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })
  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return
    timer = setTimeout(()=>{
      post({
        source: id,
        data: data.value,
        timestamp: Date.now(),
      })
    }, 10)
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}
export default data
"
`;
